packages/compiler/src/render3/view/template.ts

```diff
@@ -81,14 +81,32 @@ export function prepareEventListenerParameters(
       scope, implicitReceiverExpr, handler, 'b', eventAst.handlerSpan, implicitReceiverAccesses,
       EVENT_BINDING_SCOPE_GLOBALS);
   const statements = [];
-  if (scope) {
+  const variableDeclarations = scope?.variableDeclarations();
+  const restoreViewStatement = scope?.restoreViewStatement();
+
+  if (variableDeclarations) {
     // `variableDeclarations` needs to run first, because
     // `restoreViewStatement` depends on the result.
-    statements.push(...scope.variableDeclarations());
-    statements.unshift(...scope.restoreViewStatement());
+    statements.push(...variableDeclarations);
   }
+
   statements.push(...bindingStatements);
 
+  if (restoreViewStatement) {
+    statements.unshift(restoreViewStatement);
+
+    // If there's a `restoreView` call, we need to reset the view at the end of the listener
+    // in order to avoid a leak. If there's a `return` statement already, we wrap it in the
+    // call, e.g. `return resetView(ctx.foo())`. Otherwise we add the call as the last statement.
+    const lastStatement = statements[statements.length - 1];
+    if (lastStatement instanceof o.ReturnStatement) {
+      statements[statements.length - 1] =
+          new o.ReturnStatement(invokeInstruction(null, R3.resetView, [lastStatement.value]));
```

Comment:
Should this inherit the source span of the original statement?

```suggestion
          new o.ReturnStatement(invokeInstruction(lastStatement.sourceSpan, R3.resetView, [lastStatement.value]));
```

Reply:
Changed, but I think that we actually want `lastStatement.value.sourceSpan`.